src/Hl7.Fhir.Support.Poco/Serialization/PathStack.cs

```diff
@@ -18,66 +19,66 @@ namespace Hl7.Fhir.Serialization
     /// </summary>
     internal class PathStack
     {
-        private const string RESOURCEPREFIX = "資";
+        private const char RESOURCEPREFIX = '資';
+        private string _path = string.Empty;
 
-        private readonly Stack<string> _pathStack = new();
+        private readonly Stack<int> _tops = new();
 
         public void EnterResource(string name)
         {
-            _pathStack.Push(RESOURCEPREFIX + name);
+            _tops.Push(_path.Length);
+            _path += RESOURCEPREFIX + name;
         }
 
         public void ExitResource()
         {
-            if (_pathStack.Count == 0)
+            if (_tops.Count == 0)
                 throw new InvalidOperationException("No resource or path parts are present on the stack.");
 
-            var top = _pathStack.Pop();
-            if (top[0] != RESOURCEPREFIX[0])
+            var top = _tops.Pop();
+
+            if (_path[top] != RESOURCEPREFIX)
             {
-                _pathStack.Push(top);
+                _tops.Push(top);
                 throw new InvalidOperationException("Cannot exit a resource while inside a property.");
             }
+            else
+                _path = _path.Substring(0, top);
         }
 
         public void EnterElement(string name)
         {
-            _pathStack.Push(name);
+            bool empty = !_tops.Any();
+            _tops.Push(_path.Length);
+            _path += empty ? name : $".{name}";
         }
 
         public void ExitElement()
         {
-            if (_pathStack.Count == 0)
+            if (_tops.Count == 0)
                 throw new InvalidOperationException("No resource or path parts are present on the stack.");
 
-            var top = _pathStack.Pop();
-            if (top[0] == RESOURCEPREFIX[0])
+            var top = _tops.Pop();
+            if (_path[top] == RESOURCEPREFIX)
             {
-                _pathStack.Push(top);
+                _tops.Push(top);
                 throw new InvalidOperationException("Cannot exit a property while inside a resource.");
             }
+            else
+                _path = _path.Substring(0, top);
         }
 
         /// <summary>
         /// Return the path. Note: in contained resources, this is just the path within the contained resource.
         /// </summary>
         public string GetPath()
         {
-            if (_pathStack.Count == 0)
-                throw new InvalidOperationException("No path present.");
+            if (_tops.Count == 0) return "$this";
```

Comment:
Should this not be an empty string? Stack is empty, then there is no path, right?

Reply:
It's actually not clear what the path is.  Is it a path into a definition?  Is it a path into an instance?  I decided it would be the latter, and then it made sense to make it a FhirPath path (just like in OperationOutcome.issue).  Therefore, when you  are still in the root of an object, `$this` makes sense, more than an empty string (which debugs and display rather nastily).  

The *only* time this happens is when you validate an Element (not a resource), and immediately catch an error before going into the Element's properties.  The error will now display "At $this." instead of "At.".